# SmartPrix Laptop Dataset Analysis

## Technical Specifications and User Experience Analysis

Welcome to the `Technical Specifications and User Experience analysis` notebook! 

In this notebook, we're getting into the nitty-gritty of laptop performance and user satisfaction. We'll be looking at how different specs like processors, RAM, and storage options correlate with user ratings and overall performance scores. 

We will analyze **performance metrics, physical characteristics, and user ratings** to gain a comprehensive understanding of what features contribute to overall satisfaction and how they vary across different models.

We'll also dive into some interesting questions about *physical characteristics like weight and screen size, hardware features, connectivity* and how they relate to user preferences. 

The goal is to understand what **technical aspects really matter to users** and how that translates into their overall experience. 

As we go through the analysis, feel free to run these queries and see what insights we can draw from the data. 

Let's get started!

### Connecting to the Database
The first line of code connects us to the `smartprix_laptop_db` database where the `laptops` table resides. 
We will also connect to the `laptop_schema` schema, allowing us to start running the queries and exploring the dataset interactively.

In [54]:
import sqlalchemy
%load_ext sql
%sql postgresql:///smartprix_laptop_db
%sql SET search_path TO laptop_schema;

In [2]:
%config SqlMagic.displaycon = False
%config SqlMagic.displaylimit = None

displaylimit: Value None will be treated as 0 (no limit)

### Q1: Average spec score across different processor brands
How does the average spec score vary across different processor brands?

In [52]:
%%sql
SELECT brand, 
    COUNT(*) AS laptop_count, 
    ROUND(AVG(spec_score), 2) 
        AS avg_spec_score
FROM laptops
GROUP BY brand
ORDER BY avg_spec_score DESC;

31 rows affected.

+-----------+--------------+----------------+
|   brand   | laptop_count | avg_spec_score |
+-----------+--------------+----------------+
|   Razer   |      1       |     83.00      |
|  Ninkear  |      1       |     73.00      |
|    MSI    |     107      |     70.71      |
|     LG    |      5       |     70.20      |
|  Colorful |      1       |     70.00      |
|  Gigabyte |      6       |     69.17      |
|  Fujitsu  |      3       |     64.00      |
|  Samsung  |      21      |     63.10      |
|    Dell   |     105      |     62.24      |
|     HP    |     195      |     61.99      |
|   Xiaomi  |      6       |     61.83      |
|    Asus   |     178      |     61.43      |
| Zebronics |      8       |     60.25      |
|   Lenovo  |     202      |     59.37      |
|   Huawei  |      6       |     59.17      |
|    Acer   |      82      |     57.41      |
|   Tecno   |      1       |     56.00      |
|   Realme  |      1       |     56.00      |
|   Honor   |      3       |     55.00      |
| Microsoft |      2       |     54.50      |
|   Apple   |      17      |     53.94      |
|  Infinix  |      21      |     52.48      |
|   Wings   |      4       |     49.00      |
|   Chuwi   |      8       |     45.63      |
|  Ultimus  |      4       |     42.75      |
|   Avita   |      2       |     42.00      |
|   iBall   |      1       |     35.00      |
|    AXL    |      2       |     33.50      |
|   Walker  |      1       |     33.00      |
| Primebook |      2       |     30.00      |
|    Jio    |      2       |     21.00      |
+-----------+--------------+----------------+

**Razer** leads with an `average spec score` of 83.00 *though it has only 1 laptop in the dataset*, followed by high-performance brands like **MSI (70.71 with 107 laptops)** and **LG (70.20 with 5 laptops)**.

Popular brands like **Dell, HP, Asus and Lenovo** have `middling average spec score` **(around 60)** with **more laptops** in the dataset. 

Some lesser-known brands like Jio and Primebook have lower average spec scores (21.00 and 30.00 respectively). 

This suggests a clear differentiation in performance capabilities across various laptop manufacturers.

### Q2: Percentage of laptops with SSD storage
What percentage of laptops in the dataset have SSD storage?

In [51]:
%%sql 
SELECT CONCAT(
  ROUND(100.0 * 
    SUM(CASE WHEN internal_memory LIKE '%SSD%' 
        THEN 1 ELSE 0 END) 
        / COUNT(*), 2), 
        '%') 
    AS pct_ssd_laptops
FROM laptops;

1 rows affected.

+-----------------+
| pct_ssd_laptops |
+-----------------+
|      98.80%     |
+-----------------+

### Q3: Distribution of processor cores across laptop utilities
How does the distribution of processor cores vary across different laptop utilities (e.g., business, gaming, everyday use)?

In [50]:
%%sql
SELECT utility, 
    processor_core, 
    COUNT(*) AS laptop_count
FROM laptops
WHERE utility IS NOT NULL AND processor_core IS NOT NULL
GROUP BY utility, processor_core
ORDER BY utility DESC, 
    laptop_count DESC;

43 rows affected.

+---------------------------------------------+----------------+--------------+
|                   utility                   | processor_core | laptop_count |
+---------------------------------------------+----------------+--------------+
|          Performance, Everyday Use          |    Core i5     |      7       |
|          Performance, Everyday Use          |    Core i3     |      2       |
|     Performance, Business, Everyday Use     |    Core i7     |      1       |
|            Performance, Business            |    Core i7     |      3       |
|                 Performance                 |    Core i5     |     125      |
|                 Performance                 |    Core i7     |      79      |
|                 Performance                 |    Core i3     |      63      |
|                 Performance                 |    Core i9     |      20      |
|                 Performance                 |   Core N4500   |      1       |
|             Gaming, Performance             |    Core i5     |      9       |
|             Gaming, Performance             |    Core i9     |      5       |
|             Gaming, Performance             |    Core i7     |      4       |
|             Gaming, Everyday Use            |    Core i7     |      1       |
|                    Gaming                   |    Core i5     |      73      |
|                    Gaming                   |    Core i7     |      33      |
|                    Gaming                   |    Core i9     |      20      |
|          Everyday Use, Performance          |    Core i5     |      12      |
|          Everyday Use, Performance          |    Core i3     |      7       |
|          Everyday Use, Performance          |    Core i7     |      4       |
|          Everyday Use, Performance          |    Core i9     |      1       |
| Everyday Use, Gaming, Business, Performance |    Core i5     |      5       |
| Everyday Use, Gaming, Business, Performance |    Core i3     |      1       |
| Everyday Use, Gaming, Business, Performance |    Core i7     |      1       |
|             Everyday Use, Gaming            |    Core i7     |      2       |
| Everyday Use, Business, Performance, Gaming |    Core i5     |      3       |
| Everyday Use, Business, Performance, Gaming |    Core i3     |      2       |
|     Everyday Use, Business, Performance     |    Core i5     |      17      |
|     Everyday Use, Business, Performance     |    Core i3     |      14      |
|     Everyday Use, Business, Performance     |    Core i7     |      6       |
|            Everyday Use, Business           |    Core i5     |      3       |
|            Everyday Use, Business           |    Core i3     |      1       |
|                 Everyday Use                |    Core i5     |      49      |
|                 Everyday Use                |    Core i3     |      16      |
|                 Everyday Use                |    Core i7     |      15      |
|                 Everyday Use                |    Core i9     |      4       |
|                 Everyday Use                |   Core Z3735   |      1       |
|            Business, Performance            |    Core i5     |      15      |
|            Business, Performance            |    Core i7     |      8       |
|            Business, Performance            |    Core i3     |      5       |
|            Business, Performance            |    Core i9     |      2       |
|                   Business                  |    Core i5     |      7       |
|                   Business                  |    Core i7     |      6       |
|                   Business                  |    Core i3     |      3       |
+---------------------------------------------+----------------+--------------+

The data shows that `Core i5 processors` are the **most popular** in almost all utility categories, especially for `"Performance and Gaming"` laptops. 

`Core i7` and `Core i9 processors` are notably **more frequent** in **high-performance** categories like `"Gaming"` and `"Performance"`. 

While `Core i3 CPUs` are **more common** in `"Everyday Use"` and cost-effective configurations.

### Q4: Distribution of Intel vs. AMD processors
What is the distribution of laptops with Intel vs. AMD processors?

In [49]:
%%sql
SELECT processor_brand, 
    COUNT(*) AS laptop_count
FROM laptops
WHERE processor_brand IS NOT NULL
GROUP BY processor_brand
ORDER BY laptop_count DESC;

2 rows affected.

+-----------------+--------------+
| processor_brand | laptop_count |
+-----------------+--------------+
|      Intel      |     743      |
|       Amd       |     225      |
+-----------------+--------------+

With **more than three times** as many laptops with `Intel processors (743)` as `AMD processors (225)`, the data clearly shows that Intel processors dominate the laptop market in this dataset.

### Q5: Types and frequency of ROM memory
What are the different types of ROM memory available in the laptops, and how frequently do they occur?

In [48]:
%%sql
SELECT rom_memory, 
    COUNT(*) AS laptop_count
FROM laptops
GROUP BY rom_memory
ORDER BY laptop_count DESC;

42 rows affected.

+-------------------+--------------+
|     rom_memory    | laptop_count |
+-------------------+--------------+
|   16 GB DDR4 RAM  |     247      |
|   8 GB DDR4 RAM   |     237      |
|   16 GB DDR5 RAM  |     136      |
|  16 GB LPDDR5 RAM |     106      |
|  8 GB LPDDR5 RAM  |      49      |
|   32 GB DDR5 RAM  |      42      |
|   8 GB DDR5 RAM   |      23      |
| 16 GB LPDDR4X RAM |      20      |
| 16 GB LPDDR5X RAM |      18      |
|  8 GB LPDDR4X RAM |      13      |
|     16 GB RAM     |      9       |
|  32 GB LPDDR5 RAM |      9       |
|      8 GB RAM     |      9       |
|   4 GB DDR4 RAM   |      8       |
|  8 GB LPDDR4 RAM  |      7       |
| 32 GB LPDDR5X RAM |      7       |
|  4 GB LPDDR4X RAM |      7       |
| 32 GB LPDDR5x RAM |      6       |
| 16 GB LPDDR5x RAM |      5       |
|  4 GB LPDDR4 RAM  |      4       |
|   32 GB DDR4 RAM  |      3       |
| 16 GB LPDDR4x RAM |      3       |
|   64 GB DDR5 RAM  |      3       |
|     18 GB RAM     |      3       |
|  16 GB LPDDR4 RAM |      3       |
|  8 GB LPDDR4x RAM |      2       |
|     36 GB RAM     |      2       |
| 32 GB LPDDR4X RAM |      2       |
|  16 GB DDR5 SDRAM |      2       |
|  64 GB LPDDR5 RAM |      1       |
| 16 GB LPDDRx4 RAM |      1       |
|  16 GB PDDR5X RAM |      1       |
|  4 GB ‎LPDDR4 RAM  |      1       |
|   12 GB DDR4 RAM  |      1       |
|  12 GB LPDDR3 RAM |      1       |
|   8 GB DDR3 RAM   |      1       |
|   16 GB DDR6 RAM  |      1       |
|     48 GB RAM     |      1       |
|   2 GB DDR3 RAM   |      1       |
|      DDR5 RAM     |      1       |
|  8 GB DDR5 SDRAM  |      1       |
|  12 GB LPDDR4 RAM |      1       |
+-------------------+--------------+

The data reveals a **wide variety of RAM configurations** in laptops, with DDR4 and DDR5 being the most common types and 16 GB and 8 GB being the most frequent capacities.

In particular, **16 GB DDR4 RAM** is the **most common type (247 laptops)**, closely followed by **8 GB DDR4 RAM (237 laptops)** and **16 GB DDR5 RAM (136 laptops)**.

Although DDR4 and DDR5 are the most common, LPDDR5 RAM is also widely used, especially in laptops with a 16 GB capacity (106 laptops).

There's a wide range of laptop memory configurations to meet different user needs and budgets, including high-capacity options like 32 GB and 64 GB, as well as less typical configurations like LPDDR4X, LPDDR5X, and even a few instances of older DDR3 RAM.

### Q6: Common graphics card brands in high-rated laptops
Which graphics card brands are most commonly used in high-rated laptops?

In [47]:
%%sql

-- define high-rated laptops as those that fall within 
-- the top 10% of their spec_score values

-- Identify the laptops in the 90th percentile of spec_score

WITH spec_score_distribution AS 
(
SELECT spec_score, 
    NTILE(10) OVER (ORDER BY spec_score DESC) AS percentile_rank
FROM laptops
)
SELECT graphics_card, 
    COUNT(*) AS laptop_count
FROM laptops
WHERE spec_score IN (
    SELECT spec_score 
    FROM spec_score_distribution 
    WHERE percentile_rank = 1
)
GROUP BY graphics_card
ORDER BY laptop_count DESC;

20 rows affected.

+--------------------------------------------+--------------+
|               graphics_card                | laptop_count |
+--------------------------------------------+--------------+
|   8 Gb Nvidia Geforce Rtx 4060 Graphics    |      44      |
|   8 Gb Nvidia Geforce Rtx 4070 Graphics    |      18      |
|   6 Gb Nvidia Geforce Rtx 4050 Graphics    |      12      |
|   12 Gb Nvidia Geforce Rtx 4080 Graphics   |      11      |
|   16 Gb Nvidia Geforce Rtx 4090 Graphics   |      10      |
|             Intel Arc Graphics             |      3       |
|   4 Gb Nvidia Geforce Rtx 3050 Graphics    |      2       |
|       6 Gb Nvidia Rtx 4050 Graphics        |      2       |
|   12 Gb Nvidia Geforce Rtx4080 Graphics    |      1       |
|   6 Gb Nvidia Geforce Rtx 3060 Graphics    |      1       |
| 8 Gb Nvidia Geforce Rtx A2000 Ada Graphics |      1       |
| 16 Gb Nvidia Geforce Rtx 3080 Ti Graphics  |      1       |
| 12 Gb Nvidia Geforce Rtx 3500 Ada Graphics |      1       |
|  8 Gb Nvidia Geforce Rtx 3070 Ti Graphics  |      1       |
|     8 Gb Amd Radeon Rx 7600s Graphics      |      1       |
| 12 Gb Nvidia Geforce Rtx 3080 Ti Graphics  |      1       |
|     4 Gb Nvidia Quadro T1200 Graphics      |      1       |
|            Amd Radeon Graphics             |      1       |
|           Intel Iris Xe Graphics           |      1       |
|       4 Gb Intel Arc A370m Graphics        |      1       |
+--------------------------------------------+--------------+

**NVIDIA graphics cards** are clearly the most common in high-rated laptops, as evidenced by the **8GB Nvidia GeForce RTX 4060**, which is present in **44 laptops**, followed by the **8GB Nvidia GeForce RTX 4070** in **18 laptops**, and the **6GB Nvidia GeForce RTX 4050** in **12 laptops**.

**High-end NVIDIA cards** like the **RTX 4080 (12GB)** and **RTX 4090 (16GB)** are also well-represented, appearing in **11 and 10 laptops respectively**, indicating a significant presence of top-tier gaming laptops in the dataset.

While NVIDIA clearly dominates the list, there are a few other brands that are slightly present, such as AMD Radeon (2 laptops), Intel Arc Graphics (3 laptops), and Intel Iris Xe Graphics (1 laptop). 

These findings suggest that NVIDIA is the brand of choice for high-performance graphics in this laptops dataset.

### Q7: Common RAM and storage combinations across processor generations
What is the most common combination of RAM and storage capacity, and how has this changed across different processor generations?

In [46]:
%%sql
WITH memory_cte AS 
(
SELECT processor_gen 
    AS proc_gen, 
    CONCAT(
      TRIM(REGEXP_SUBSTR(rom_memory, '([0-9]+\s?GB)')), 
        ' + ',
      TRIM(REGEXP_SUBSTR(internal_memory, '([0-9]+\s?(GB|TB)\s?SSD)'))
-- use CONCAT, TRIM and REGEXP_SUBSTR function to
-- extract, then combine rom_memory and internal memory
-- example: if rom_memory is "8 GB LPDDR5 RAM" 
-- and internal_memory is "1 TB SSD"
-- Result: "8 GB + 1 TB SSD"
    ) AS ram_storage_combo
FROM laptops
WHERE processor_gen LIKE '%_Gen%'
    OR processor_gen LIKE 'M%' -- for apple M series
),ranked_combos AS 
(
SELECT proc_gen, 
    ram_storage_combo, 
    COUNT(*) AS laptop_count,
    RANK() OVER (PARTITION BY proc_gen 
        ORDER BY COUNT(*) DESC) 
    AS rank_in_gen,
    RANK() OVER (ORDER BY COUNT(*) DESC) 
    AS overall_rank
  FROM memory_cte
  GROUP BY proc_gen, ram_storage_combo
)
SELECT 
  proc_gen, 
  ram_storage_combo, 
  laptop_count,
  rank_in_gen,
  overall_rank
FROM ranked_combos
WHERE rank_in_gen = 1 
    OR overall_rank <= 5
ORDER BY 
    REGEXP_SUBSTR(proc_gen, 'M'), 
    REGEXP_SUBSTR(proc_gen, '[0-9]+')::INTEGER, 
    laptop_count DESC;

19 rows affected.

+----------+--------------------+--------------+-------------+--------------+
| proc_gen | ram_storage_combo  | laptop_count | rank_in_gen | overall_rank |
+----------+--------------------+--------------+-------------+--------------+
|    M1    | 8 GB + 256 GB SSD  |      1       |      1      |      45      |
|    M2    | 8 GB + 512 GB SSD  |      1       |      1      |      45      |
|    M2    | 8 GB + 256 GB SSD  |      1       |      1      |      45      |
|    M3    | 8 GB + 512 GB SSD  |      3       |      1      |      27      |
|  M3 Pro  | 18 GB + 512 GB SSD |      2       |      1      |      33      |
|  M3 Max  |  36 GB + 1 TB SSD  |      2       |      1      |      33      |
| 3rd Gen  | 8 GB + 512 GB SSD  |      8       |      1      |      16      |
| 4th Gen  | 8 GB + 256 GB SSD  |      3       |      1      |      27      |
| 5th Gen  | 8 GB + 512 GB SSD  |      33      |      1      |      10      |
| 6th Gen  | 16 GB + 512 GB SSD |      8       |      1      |      16      |
| 7th Gen  | 16 GB + 512 GB SSD |      50      |      1      |      6       |
| 8th Gen  |  16 GB + 1 TB SSD  |      2       |      1      |      33      |
| 10th Gen | 8 GB + 512 GB SSD  |      4       |      1      |      26      |
| 11th Gen | 8 GB + 512 GB SSD  |      55      |      1      |      5       |
| 12th Gen | 16 GB + 512 GB SSD |     102      |      1      |      2       |
| 12th Gen | 8 GB + 512 GB SSD  |      81      |      2      |      3       |
| 13th Gen | 16 GB + 512 GB SSD |     128      |      1      |      1       |
| 13th Gen |  16 GB + 1 TB SSD  |      73      |      2      |      4       |
| 14th Gen |  32 GB + 2 TB SSD  |      5       |      1      |      22      |
+----------+--------------------+--------------+-------------+--------------+

The data shows a **clear trend of increasing RAM and storage capacities** across `processor generations`, with the most common combination evolving from `8 GB RAM + 512 GB SSD` in earlier generations to `16 GB RAM + 512 GB SSD` in more recent ones. 

The `13th generation processors` show the highest frequency of laptops, with **128 models** featuring `16 GB RAM + 512 GB SSD` (ranked 1st overall), followed by **73 models** with `16 GB RAM + 1 TB SSD`, indicating a shift towards higher storage capacities in premium models.

Interestingly, the latest `14th generation processors` show a **significant jump in both RAM and storage**, with the most common configuration being `32 GB RAM + 2 TB SSD`, suggesting a trend towards even higher-spec machines in the newest generation, although the sample size appears smaller for this latest generation.

**Apple MacBooks (M-series processors):**

For **M1 and M2**, the most common combinations are ``"8 GB + 256 GB SSD"`` and `"8 GB + 512 GB SSD"`. 

**M3** shows a slight increase with `"8 GB + 512 GB SSD"` as the most common. 

**M3 Pro and M3 Max** show higher capacities: `"18 GB + 512 GB SSD"` and `"36 GB + 1 TB SSD"` respectively, reflecting their positioning as higher-end models.

### Q8: Average number of USB ports in different price ranges
What is the average number of USB ports for laptops in different price ranges?

In [53]:
%%sql
WITH usb_ports_cte AS
(
SELECT id, price, usb_ports,
(
SELECT SUM(CAST(num AS INTEGER))
    FROM 
-- use REGEXP_SUBSTR, TRIM, UNNEST, STRING_TO_ARRAY
-- REPLACE functions in a subquery
-- to extract the number of usb_port
-- example: "1 x USB 2.0, 2 x USB 3.0, 1 x USB Type-C"
-- becomes: 4 (total_usb_ports)
    (
    SELECT REGEXP_SUBSTR(TRIM(part), '^[0-9]+') 
        AS num
    FROM UNNEST(STRING_TO_ARRAY
        (REPLACE(usb_ports, ' x ', ','), ',')) 
        AS part
    ) AS numbers
        WHERE num != ''
    ) AS total_usb_ports
FROM laptops
)
SELECT lps.price_segment 
    AS price_segment, 
    COUNT(*) AS laptop_count, 
    ROUND(AVG(usb.total_usb_ports), 2) 
    AS avg_usb_ports
FROM laptop_price_segments lps -- from laptop_price_segments view
JOIN usb_ports_cte usb 
ON lps.id = usb.id
GROUP BY lps.price_segment;

3 rows affected.

+---------------+--------------+---------------+
| price_segment | laptop_count | avg_usb_ports |
+---------------+--------------+---------------+
|     Budget    |     201      |      3.25     |
|   Mid-range   |     565      |      3.41     |
|    Premium    |     232      |      3.67     |
+---------------+--------------+---------------+

The analysis reveals that the average number of USB ports varies across different price segments of laptops. 

**Budget** laptops have an average of `3.25 USB ports`, **Mid-range** laptops have `3.41`, and **Premium laptops**  lead with an average of `3.67 USB ports`. 

This indicates a trend where higher-priced laptops tend to offer more connectivity options.

### Q9: Feature-Rich High-Performance Laptops
Identify the laptops that have more than five distinct hardware features and port connections (e.g., Backlit Keyboard, Thunderbolt, etc.). 

For these laptops, determine which have at least a spec score of 90.

In [30]:
%%sql
WITH features_cte AS 
(
SELECT name, 
    price, 
    spec_score, 
    ARRAY_LENGTH(STRING_TO_ARRAY(
        CONCAT(port_connection, ', ', hardware_features), ', '), 1) 
    AS features_count
FROM laptops
)
SELECT name, price, 
    spec_score, features_count
FROM features_cte
WHERE features_count > 5 
    AND spec_score >= 90
ORDER BY spec_score DESC;

6 rows affected.

+---------------------------------------------+--------+------------+----------------+
|                     name                    | price  | spec_score | features_count |
+---------------------------------------------+--------+------------+----------------+
|   MSI Stealth 17 A13VH-055IN Gaming Laptop  | 489883 |     98     |       7        |
| MSI Vector 17 HX A14VHG-806IN Gaming Laptop | 279990 |     92     |       6        |
|  MSI CreatorPro X17 HX A13VKS-249IN Laptop  | 519990 |     92     |       7        |
|   Lenovo Legion Y9000K 2024 Gaming Laptop   | 349990 |     92     |       6        |
|       Dell Alienware m16 Gaming Laptop      | 303290 |     91     |       7        |
|    MSI Creator Z17 HX Studio A13V Laptop    | 247999 |     90     |       7        |
+---------------------------------------------+--------+------------+----------------+

According to the data, **six laptops** have a `spec score of more than 90` and `more than five different hardware features and port connections`. There are three brands represented: **MSI, Lenovo, and Dell**. MSI leads the pack with four laptops on the list.

These feature-rich, high-performance laptops **range widely in price** from *247,999 to 519,990*, indicating that although all of these top-tier models match the requirements for a large number of features and good spec scores, there is a notable difference in cost.

### Q10: Above-Average Spec Scores Within Brand and Price Range
Identify the top 5 laptops that have a higher spec score than the average spec score of laptops in their same price range and brand.

In [28]:
%%sql

-- note: I already created the laptop price segments view in the
-- market and value analysis notebook

WITH relevant_data AS 
(
SELECT lps.brand, lps.name, 
    lps.price_segment, 
    lps.price, 
    l.spec_score
FROM laptops l
JOIN laptop_price_segments lps
ON l.id = lps.id
), brand_cte AS 
(
SELECT brand, 
    ROUND(AVG(spec_score), 2) 
    AS brand_avg_spec_score
FROM relevant_data
GROUP BY brand
)
, price_segment_cte AS 
(
SELECT price_segment, 
    ROUND(AVG(spec_score), 2) 
    AS price_segment_avg_spec_score
FROM relevant_data
GROUP BY price_segment
), ranked_laptops AS 
(
SELECT 
    r.price_segment, r.brand, r.name, 
    r.price, r.spec_score, 
    b.brand_avg_spec_score, 
    p.price_segment_avg_spec_score, 
    ROW_NUMBER() OVER(PARTITION BY r.price_segment 
    ORDER BY r.spec_score DESC) 
    AS rank
FROM relevant_data r
JOIN brand_cte b 
ON r.brand = b.brand
JOIN price_segment_cte p 
ON r.price_segment = p.price_segment
WHERE r.spec_score > b.brand_avg_spec_score
AND r.spec_score > p.price_segment_avg_spec_score
)
SELECT *
FROM ranked_laptops
WHERE rank <= 5;

15 rows affected.

+---------------+-----------+-----------------------------------------------------------+--------+------------+----------------------+------------------------------+------+
| price_segment |   brand   |                            name                           | price  | spec_score | brand_avg_spec_score | price_segment_avg_spec_score | rank |
+---------------+-----------+-----------------------------------------------------------+--------+------------+----------------------+------------------------------+------+
|     Budget    | Zebronics |          Zebronics Pro Series Z ZEB-NBC 4S Laptop         | 32990  |     64     |        60.25         |            48.43             |  1   |
|     Budget    |   Lenovo  |          Lenovo ThinkBook 15 G5 21JF002JIN Laptop         | 33790  |     62     |        59.37         |            48.43             |  2   |
|     Budget    | Zebronics |       Zebronics Pro Series Z ZEB-NBC 4S 2023 Laptop       | 31990  |     62     |        60.25         |            48.43             |  3   |
|     Budget    | Zebronics |       Zebronics Pro Series Z ZEB-NBC 3S 2023 Laptop       | 27990  |     61     |        60.25         |            48.43             |  4   |
|     Budget    |  Infinix  |       Infinix INBook X2 Slim Series XL23 2023 Laptop      | 36990  |     55     |        52.48         |            48.43             |  5   |
|   Mid-range   |    Asus   |      Asus ROG Flow X13 GV301RC-LJ073WS Gaming Laptop      | 89990  |     75     |        61.43         |            60.69             |  1   |
|   Mid-range   |   Lenovo  |              Lenovo Yoga 7 82YM0074IN Laptop              | 94190  |     75     |        59.37         |            60.69             |  2   |
|   Mid-range   |    MSI    |           MSI Sword 15 A12VF-401IN Gaming Laptop          | 94990  |     74     |        70.71         |            60.69             |  3   |
|   Mid-range   |    Asus   |   Asus TUF Gaming F15 2023 FX507ZV-LP094W Gaming Laptop   | 99990  |     74     |        61.43         |            60.69             |  4   |
|   Mid-range   |    Asus   |      Asus ROG Flow X13 GV301RC-LJ132WS Gaming Laptop      | 84990  |     74     |        61.43         |            60.69             |  5   |
|    Premium    |    MSI    |          MSI Stealth 17 A13VH-055IN Gaming Laptop         | 489883 |     98     |        70.71         |            73.17             |  1   |
|    Premium    |    MSI    |   MSI Stealth 16 Mercedes AMG A13VG-264IN Gaming Laptop   | 279990 |     95     |        70.71         |            73.17             |  2   |
|    Premium    |    Dell   |                  Dell XPS 17 9730 Laptop                  | 355490 |     93     |        62.24         |            73.17             |  3   |
|    Premium    |    MSI    |        MSI Raider GE78 HX 14VHG-805IN Gaming Laptop       | 299990 |     93     |        70.71         |            73.17             |  4   |
|    Premium    |    MSI    | MSI Stealth 16 Mercedes AMG Motorsport A13V Gaming Laptop | 292990 |     92     |        70.71         |            73.17             |  5   |
+---------------+-----------+-----------------------------------------------------------+--------+------------+----------------------+------------------------------+------+

According to the results, **Zebronics** dominates the `budget laptop market`, taking three of the top five spots for laptops having a higher spec score than the average spec score in the same price segment and brand. 

**Lenovo and Infinix** each have one spot, and all of these laptops have spec scores that are significantly higher than the segment average of `48.43`. 

The Mid-range and Premium segments, on the other hand, showcase a wider variety of brands. **Asus and MSI** are well-known in the `Mid-range category`, while **MSI, Dell, and Lenovo** leads the `premium category`. 

These brands all significantly outperform their respective average spec scores of `60.69` and `73.17`.

### Q11: Average weight across screen size categories
How does the average weight of laptops differ across screen size categories?

In [31]:
%%sql
WITH screen_size_cte AS 
(
SELECT 
    CAST(REGEXP_SUBSTR(screen_size, '[0-9]+(\.[0-9]+)?') 
         AS DECIMAL(4,1)) 
    AS screen_size_inches,
    weight
FROM 
    laptops
)
SELECT 
    CASE 
     WHEN screen_size_inches <= 14 
        THEN '14" and below' 
     WHEN screen_size_inches BETWEEN 14.1 AND 15.9 
        THEN '14" - 15"' 
     ELSE '16" and above' END 
    AS screen_size_category,
    ROUND(AVG(weight), 2) AS avg_weight
FROM screen_size_cte
GROUP BY CASE WHEN screen_size_inches <= 14 THEN '14" and below'
            WHEN screen_size_inches BETWEEN 14.1 AND 15.9 THEN '14" - 15"'
            ELSE '16" and above' END
ORDER BY avg_weight DESC;

3 rows affected.

+----------------------+------------+
| screen_size_category | avg_weight |
+----------------------+------------+
|    16" and above     |    2.20    |
|      14" - 15"       |    1.90    |
|    14" and below     |    1.39    |
+----------------------+------------+

The analysis of average weight across different screen size categories shows that laptops with a **screen size of 16 inches and above** have the **highest average weight** at `2.20 kg`. 

In contrast, laptops that are **14 inches and below** are the **lightest**, with an average weight of `1.39 kg`, *indicating that larger laptops tend to be heavier.*

### Q12: Brands with a Wide Range of Screen Sizes
Which brands have laptops with a wide range of screen sizes?

In [15]:
%%sql
SELECT brand, 
    MIN(screen_size_inches) AS min_screen_size, 
    MAX(screen_size_inches) AS max_screen_size, 
    (MAX(screen_size_inches) - MIN(screen_size_inches)) 
        AS screen_size_range
FROM
(
SELECT brand, 
    CAST(REGEXP_SUBSTR(screen_size, '[0-9]+(\.[0-9]+)?') 
         AS DECIMAL(4,1)) 
    AS screen_size_inches
    FROM laptops
) AS brand_screen_size
GROUP BY brand
-- add a HAVING clause to filter out brands with 
-- only one screen size (where the range is 0).
HAVING COUNT(DISTINCT screen_size_inches) > 1  
ORDER BY screen_size_range DESC;

18 rows affected.

+----------+-----------------+-----------------+-------------------+
|  brand   | min_screen_size | max_screen_size | screen_size_range |
+----------+-----------------+-----------------+-------------------+
|   Asus   |       11.6      |       17.3      |        5.7        |
|   Dell   |       13.3      |       18.0      |        4.7        |
|   MSI    |       13.3      |       18.0      |        4.7        |
|  Lenovo  |       11.6      |       16.0      |        4.4        |
|   Acer   |       11.6      |       16.0      |        4.4        |
|    HP    |       13.3      |       17.3      |        4.0        |
|  Chuwi   |       12.0      |       15.6      |        3.6        |
|  Huawei  |       12.6      |       16.0      |        3.4        |
|    LG    |       14.0      |       17.0      |        3.0        |
|  Apple   |       13.3      |       16.2      |        2.9        |
| Samsung  |       13.3      |       16.0      |        2.7        |
|  Xiaomi  |       14.0      |       16.1      |        2.1        |
|  Honor   |       14.0      |       16.0      |        2.0        |
| Infinix  |       14.0      |       16.0      |        2.0        |
|  Wings   |       14.0      |       15.6      |        1.6        |
|   AXL    |       14.1      |       15.6      |        1.5        |
| Gigabyte |       15.0      |       16.0      |        1.0        |
| Fujitsu  |       13.3      |       14.0      |        0.7        |
+----------+-----------------+-----------------+-------------------+

The output shows that **Asus** has the **widest range of screen sizes** among laptop brands, spanning from `11.6 to 17.3 inches`, resulting in a `5.7-inch range`. 

**Dell and MSI** follow closely, both offering screens from `13.3 to 18.0 inches (4.7-inch range)`. Other major brands like **Lenovo, Acer, and HP** also demonstrate significant variety in their screen size offerings, with ranges between `4.0 and 4.4 inches`.

A few of the brands that are not included in the results provide only one screen size choice.

### Q13: Weight-to-Screen Size Ratio Analysis
Calculate the ratio of weight to screen size for all laptops. 

Identify the top 10 laptops with the lowest ratio (lightest for their screen size) and the bottom 10 (heaviest for their screen size). 

How do these laptops compare in terms of performance and price?

In [14]:
%%sql
WITH relevant_data AS 
(
SELECT name, 
    spec_score, 
    price,
    ROUND(1.0 * weight / 
        CAST(REGEXP_SUBSTR(screen_size, '[0-9]+(\.[0-9]+)?') 
             AS DECIMAL(3,1)), 2) 
    AS ws_ratio
FROM laptops
WHERE weight IS NOT NULL 
    AND screen_size IS NOT NULL
),lightest_laptops AS 
(
SELECT 'Lightest laptop' 
    AS weight_screen_category, 
    name, 
    ws_ratio, 
    price, 
    spec_score
FROM relevant_data
ORDER BY ws_ratio ASC
LIMIT 10
),heaviest_laptops AS 
(
SELECT 'Heaviest laptop' 
    AS weight_screen_category, 
    name, 
    ws_ratio, 
    price, 
    spec_score
FROM relevant_data
ORDER BY ws_ratio DESC
LIMIT 10
)
SELECT *, 
    ROUND(AVG(spec_score) OVER(), 2) 
    AS avg_spec_score,
    ROUND(AVG(price) OVER(), 2) 
    AS avg_price 
FROM lightest_laptops
UNION ALL
SELECT *, 
    ROUND(AVG(spec_score) OVER(), 2) 
    AS avg_spec_score, 
    ROUND(AVG(price) OVER(), 2) 
    AS avg_price 
FROM heaviest_laptops;

20 rows affected.

+------------------------+------------------------------------------------+----------+--------+------------+----------------+-----------+
| weight_screen_category |                      name                      | ws_ratio | price  | spec_score | avg_spec_score | avg_price |
+------------------------+------------------------------------------------+----------+--------+------------+----------------+-----------+
|    Lightest laptop     |         Huawei MateBook E 2023 Laptop          |   0.06   | 88999  |     53     |     62.30      | 103858.30 |
|    Lightest laptop     |       MSI Prestige 13 AI Evo A1MG Laptop       |   0.07   | 99990  |     66     |     62.30      | 103858.30 |
|    Lightest laptop     |         Fujitsu UH-X ‎4ZR1L12856 Laptop         |   0.07   | 89990  |     69     |     62.30      | 103858.30 |
|    Lightest laptop     |       MSI Prestige 13 AI Evo A1M Laptop        |   0.07   | 113999 |     64     |     62.30      | 103858.30 |
|    Lightest laptop     |       Samsung Galaxy Book2 Pro 13 Laptop       |   0.07   | 62990  |     60     |     62.30      | 103858.30 |
|    Lightest laptop     |              Chuwi Ubook X Laptop              |   0.07   | 39990  |     46     |     62.30      | 103858.30 |
|    Lightest laptop     |             HP Dragonfly G4 Laptop             |   0.07   | 231746 |     69     |     62.30      | 103858.30 |
|    Lightest laptop     |     MSI Prestige 13 Evo A12M-085IN Laptop      |   0.07   | 77990  |     61     |     62.30      | 103858.30 |
|    Lightest laptop     |    MSI Prestige 13 AI Evo A1MG 2024 Laptop     |   0.07   | 114990 |     69     |     62.30      | 103858.30 |
|    Lightest laptop     |     LG Gram 14 2023 ‎14Z90R-G.CH75A2 Laptop     |   0.07   | 117899 |     66     |     62.30      | 103858.30 |
|    Heaviest laptop     |      Dell Alienware M18 R1 Gaming Laptop       |   0.21   | 366190 |     84     |     77.50      | 194420.00 |
|    Heaviest laptop     |        Dell Alienware m16 Gaming Laptop        |   0.20   | 220990 |     80     |     77.50      | 194420.00 |
|    Heaviest laptop     |        Dell Alienware m16 Gaming Laptop        |   0.20   | 303290 |     91     |     77.50      | 194420.00 |
|    Heaviest laptop     |     Dell Alienware m16 2023 Gaming Laptop      |   0.20   | 199290 |     81     |     77.50      | 194420.00 |
|    Heaviest laptop     |          Dell ‎G16 7630 Gaming Laptop           |   0.19   | 187490 |     77     |     77.50      | 194420.00 |
|    Heaviest laptop     |          Dell ‎G16-7630 Gaming Laptop           |   0.19   | 178490 |     75     |     77.50      | 194420.00 |
|    Heaviest laptop     |        Dell 2023 ‎G16-7630 Gaming Laptop        |   0.18   | 166490 |     78     |     77.50      | 194420.00 |
|    Heaviest laptop     | Dell G15-5530 GN5530D83M6001ORB1 Gaming Laptop |   0.18   | 79990  |     66     |     77.50      | 194420.00 |
|    Heaviest laptop     |        Dell ‎G16-7630 2023 Gaming Laptop        |   0.18   | 139990 |     74     |     77.50      | 194420.00 |
|    Heaviest laptop     | Dell G15-5530 GN5530194YM001ORB1 Gaming Laptop |   0.18   | 101990 |     69     |     77.50      | 194420.00 |
+------------------------+------------------------------------------------+----------+--------+------------+----------------+-----------+

Based on their **weight-to-screen ratio**, the output lists the `top 10 lightest and heaviest laptops` along with their costs, performance ratings, and category averages. 

The **Huawei MateBook E 2023** leads the pack of the `lightest laptops`, with ratios between **0.06 and 0.07**, an **average spec score of 62.30**, and an **average cost of 103,858.30**. 

The **Dell Alienware M18 R1 Gaming Laptop** leads the group of `heaviest laptops`, which have ratios between **0.18 and 0.21**. 

These `heaviest laptops` also have **higher average spec scores (77.50)** and significantly **higher average prices (194,420.00)**, suggesting that heavier laptops tend to be *more powerful and expensive gaming machines.*

### Q14: Highest user-rated laptops and their price comparison
Which laptops have the highest user rating, and how does their price compare to the average?

In [33]:
%%sql
WITH highest_rated_laptops AS 
(
SELECT name, user_rating, price
FROM laptops 
WHERE user_rating = (SELECT MAX(user_rating) FROM laptops)
), average_cte AS 
(
SELECT ROUND(AVG(price), 2) 
    AS avg_price
FROM laptops
)
SELECT name, user_rating, 
    price, avg_price, 
    (price - avg_price) AS price_diff, 
    ROUND(100.0 * (price - avg_price)
          /(avg_price), 2) 
    AS percent_diff
FROM highest_rated_laptops
CROSS JOIN average_cte
ORDER BY percent_diff DESC;

31 rows affected.

+-----------------------------------------------------+-------------+--------+-----------+------------+--------------+
|                         name                        | user_rating | price  | avg_price | price_diff | percent_diff |
+-----------------------------------------------------+-------------+--------+-----------+------------+--------------+
|           Apple MacBook Pro 14 2023 Laptop          |     4.8     | 186899 |  84819.22 | 102079.78  |    120.35    |
|  Samsung Galaxy Book 4 Pro 14 NP940XGK-KG3IN Laptop |     4.8     | 171990 |  84819.22 |  87170.78  |    102.77    |
|        MSI Prestige 16 AI Studio B1VEG Laptop       |     4.8     | 159990 |  84819.22 |  75170.78  |    88.62     |
|          HP Omen 16-wf0054TX Gaming Laptop          |     4.8     | 151490 |  84819.22 |  66670.78  |    78.60     |
|        LG Gram 17 2023 17Z90R-G.CH77A2 Laptop       |     4.8     | 146990 |  84819.22 |  62170.78  |    73.30     |
|  Asus Zenbook S13 OLED 2024 UX5304MA-NQ762WS Laptop |     4.8     | 139990 |  84819.22 |  55170.78  |    65.05     |
|         Acer Nitro 16 ‎AN16-41 Gaming Laptop         |     4.8     | 129990 |  84819.22 |  45170.78  |    53.26     |
|       MSI Pulse 17 B13VGK-666IN Gaming Laptop       |     4.8     | 128990 |  84819.22 |  44170.78  |    52.08     |
|            Acer Swift Edge 16 2023 Laptop           |     4.8     | 107399 |  84819.22 |  22579.78  |    26.62     |
|        MSI Bravo 15 C7VFK-087IN Gaming Laptop       |     4.8     | 104989 |  84819.22 |  20169.78  |    23.78     |
|       Lenovo IdeaPad Slim 5i 82XF003DIN Laptop      |     4.8     | 90000  |  84819.22 |  5180.78   |     6.11     |
|       Asus Vivobook 16X K3605VU-MB541WS Laptop      |     4.8     | 89990  |  84819.22 |  5170.78   |     6.10     |
|    Asus Vivobook 16X 2023 K3605ZF-MBN741WS Laptop   |     4.8     | 88990  |  84819.22 |  4170.78   |     4.92     |
|      Lenovo Yoga Slim 6 14IRH8 83E00012INLaptop     |     4.8     | 81400  |  84819.22 |  -3419.22  |    -4.03     |
|      Dell Vostro 5630 IN5630P8YRR001ORS1 Laptop     |     4.8     | 78980  |  84819.22 |  -5839.22  |    -6.88     |
|      Asus Vivobook 16X K3605ZU-MBN541WS Laptop      |     4.8     | 76990  |  84819.22 |  -7829.22  |    -9.23     |
|     Asus Vivobook Pro 15 M6500QC-LK742WS Laptop     |     4.8     | 73990  |  84819.22 | -10829.22  |    -12.77    |
|        Dell Inspiron 7430 2023 2 in 1 Laptop        |     4.8     | 69990  |  84819.22 | -14829.22  |    -17.48    |
|         HP Victus 15-fa1124TX Gaming Laptop         |     4.8     | 59997  |  84819.22 | -24822.22  |    -29.26    |
|         Xiaomi Redmi Book Pro 14 2024 Laptop        |     4.8     | 59990  |  84819.22 | -24829.22  |    -29.27    |
| Asus Vivobook Pro 15 M6500QF-HN541WS Creator Laptop |     4.8     | 57990  |  84819.22 | -26829.22  |    -31.63    |
|         HP Victus 15-fb1018AX Gaming Laptop         |     4.8     | 54990  |  84819.22 | -29829.22  |    -35.17    |
|      Acer Aspire 5 A514-56M 2023 Gaming Laptop      |     4.8     | 53990  |  84819.22 | -30829.22  |    -36.35    |
|                HP 15s-EQ2084AU Laptop               |     4.8     | 52199  |  84819.22 | -32620.22  |    -38.46    |
|                HP 15s-er2004AU Laptop               |     4.8     | 50990  |  84819.22 | -33829.22  |    -39.88    |
|       Lenovo IdeaPad Slim 3 82H803HQIN Laptop       |     4.8     | 48900  |  84819.22 | -35919.22  |    -42.35    |
|          Lenovo S14 Gen 3 82TW000VIH Laptop         |     4.8     | 45990  |  84819.22 | -38829.22  |    -45.78    |
|       Lenovo Thinkpad E16 G1 21JN004DIG Laptop      |     4.8     | 42990  |  84819.22 | -41829.22  |    -49.32    |
|      Acer Aspire 3 Spin 14 NX.KENSI.002 Laptop      |     4.8     | 39990  |  84819.22 | -44829.22  |    -52.85    |
|                HP 14s-dq5138tu Laptop               |     4.8     | 39990  |  84819.22 | -44829.22  |    -52.85    |
|       Lenovo IdeaPad Slim 3 82RK00VVIN Laptop       |     4.8     | 37

**The output provide the following three important insights:**

1. The laptops with the highest ratings, all with `4.8 user ratings`, span a wide price range, from **premium devices** like the *Apple MacBook Pro 14 2023 (186,899)* to more **budget-friendly options** like the *Lenovo IdeaPad Slim 3 (37,990)*.

2. Many of the `highest-rated laptops`, particularly those from **premium brands** like *Apple, Samsung, and MSI*, are **significantly more expensive** than the average laptop price of 84,819.22, with some costing **over 100% more** than the average.

3. There are also **several highly-rated laptops** that are **priced below the average**, with some models from brands like *HP, Acer, and Lenovo* costing **30-55% less** than the average price, suggesting that user satisfaction is not solely determined by price.

### Q15: Correlation between user votes and user rating
Is there a correlation between the number of user votes and the user rating?

In [12]:
%%sql
SELECT 
    ROUND(CORR(user_votes, user_rating)::NUMERIC, 2) 
        AS correlation
FROM laptops;

1 rows affected.

+-------------+
| correlation |
+-------------+
|    -0.08    |
+-------------+

According to the correlation output, `user ratings` and `votes` appear to have a **very weak inverse relationship**, which means that as the **number of votes increases**, the **rating tends to slightly decrease**, but this relationship is minimal and not significant.

### Q16: Average User Rating For Different Screen Sizes
How does the average user rating change for laptops with different screen sizes?

In [10]:
%%sql
WITH screen_size_cte AS 
(
SELECT 
    CAST(REGEXP_SUBSTR(screen_size, '[0-9]+(\.[0-9]+)?') 
         AS DECIMAL(4,1)) AS screen_size_inches,
    user_rating,
    ROUND(AVG(user_rating) OVER(), 2) 
    AS overall_avg
FROM 
    laptops
)
SELECT 
    CASE WHEN screen_size_inches <= 14 
            THEN '14" and below'
         WHEN screen_size_inches BETWEEN 14.1 AND 15.9 
            THEN '14" - 15"'
         ELSE '16" and above' END 
        AS screen_size_category,
    ROUND(AVG(user_rating), 2) AS avg_rating,
    COUNT(*) AS laptop_count,
    MAX(overall_avg) AS overall_avg
FROM screen_size_cte
GROUP BY 
    CASE WHEN screen_size_inches <= 14 THEN '14" and below'
         WHEN screen_size_inches BETWEEN 14.1 AND 15.9 THEN '14" - 15"'
         ELSE '16" and above' END;

3 rows affected.

+----------------------+------------+--------------+-------------+
| screen_size_category | avg_rating | laptop_count | overall_avg |
+----------------------+------------+--------------+-------------+
|    14" and below     |    4.31    |     272      |     4.32    |
|      14" - 15"       |    4.29    |     517      |     4.32    |
|    16" and above     |    4.38    |     209      |     4.32    |
+----------------------+------------+--------------+-------------+

The data shows that laptops with **larger screens (16" and above)** have the **highest average user rating** of **4.38**, while mid-sized laptops (14" - 15") have the lowest at 4.29. 

Smaller laptops (14" and below) fall in between with a rating of 4.32, which matches the overall average rating across all categories.

### Q17: Battery Life Across Utility Category
What is the average battery life across different laptop utilities (e.g., gaming, business)?

In [9]:
%%sql
SELECT utility, 
  ROUND(AVG(CAST(
      TRIM(REPLACE(SPLIT_PART(battery, 'Wh', 1), ' ', '')) 
            AS DECIMAL(4, 1))), 2) 
    AS avg_battery_life
FROM laptops
WHERE battery LIKE '%Wh%' 
AND utility IS NOT NULL
GROUP BY utility
ORDER BY avg_battery_life DESC;

20 rows affected.

+---------------------------------------------+------------------+
|                   utility                   | avg_battery_life |
+---------------------------------------------+------------------+
|            Performance, Business            |      77.20       |
|            Utility: Everyday Use            |      76.00       |
|             Gaming, Performance             |      74.56       |
|             Everyday Use, Gaming            |      71.20       |
|      Everyday Use, Gaming, Performance      |      69.00       |
|                    Gaming                   |      67.87       |
|             Performance, Gaming             |      57.39       |
|            Business, Performance            |      55.08       |
|          Everyday Use, Performance          |      54.93       |
|                 Performance                 |      54.69       |
|             Gaming, Everyday Use            |      54.00       |
|                 Everyday Use                |      53.73       |
|          Performance, Everyday Use          |      52.27       |
| Everyday Use, Gaming, Business, Performance |      51.17       |
|      Everyday Use, Performance, Gaming      |      50.00       |
|            Everyday Use, Business           |      48.15       |
|                   Business                  |      47.81       |
|     Business, Everyday Use, Performance     |      45.00       |
|     Everyday Use, Business, Performance     |      43.03       |
| Everyday Use, Business, Performance, Gaming |      38.00       |
+---------------------------------------------+------------------+

Laptops categorized as `"Performance, Business"` have the **highest average battery life** at **77.20 Wh**, followed closely by `"Gaming, Performance"` laptops at **74.56 Wh**, indicating that high-performance machines often come with larger batteries. 

It's interesting to note that laptops solely categorized as `"Gaming"` have a **lower average battery life (67.87 Wh)** compared to those with multiple utility categories including gaming, *which suggests that pure gaming laptops might prioritize other features over battery capacity*.

The **lowest average battery life** is found in laptops categorized for multiple utilities including `"Everyday Use, Business, Performance, Gaming"` at **38.00 Wh**. 

This could indicate a trade-off between versatility and battery capacity in these multi-purpose devices.

### Q18: Relationship between battery capacity and Laptop weight
What is the relationship between battery capacity and laptop weight?

In [34]:
%%sql
SELECT ROUND(CORR(battery_capacity, weight)::NUMERIC, 2) 
        AS correlation
FROM
(
SELECT 
    CAST(TRIM(REPLACE(SPLIT_PART(battery, 'Wh', 1), ' ', '')) 
         AS DECIMAL(4, 1)) 
    AS battery_capacity, 
    weight
FROM laptops
WHERE battery LIKE '%Wh%' 
    AND weight IS NOT NULL
) AS battery_weight;

1 rows affected.

+-------------+
| correlation |
+-------------+
|     0.45    |
+-------------+

The **correlation coefficient** between `battery capacity` and `laptop weight` is **0.45**. 

This indicates a **moderate positive correlation** between the two variables. It suggests that as *laptop weight increases, there is a tendency for battery capacity to also increase,* but the relationship is not very strong. 

This makes sense intuitively, as larger laptops can accommodate bigger batteries, but weight and battery capacity are also influenced by other factors.